proposals/2312-matrix-uri.md

```diff
@@ -356,6 +356,95 @@ Client authors are strongly encouraged to standardise custom query elements
 that gain adoption by submitting an MSC defining them in a way compatible
 across the client ecosystem.
 
+We define four other query parameters below which improve the ability to onboard
+new matrix users (`default_client` and `default_hs`) and let link authors
+specify a mandated client or server (`force_client` and `force_hs`).
+
+##### Transparently onboarding new Matrix users
+
+New Matrix users following a matrix URI will typically expect to be onboarded
+to a specific branded client and/or server.  It is critical to minimise
+friction in doing so, in order to compete with the UX of traditional
+centralised apps. For instance, FOSDEM wants to onboard new users straight
+into https://chat.fosdem.org and default to using an account on the fosdem.org
+homeserver, via a single click.  To support this, we can specify query
+parameters to the URI to specify a default client and homeserver for new
+users:
+
+* `default_client` optionally specifies the client URI new users following this
+link should use to connect to Matrix, if there is not a registered matrix URI
+handler already.  To convert the matrix URL into the suggested default client
+URL, the caller should replace the matrix URI scheme prefix with the URL given
+by the `default_client` parameter, suffixed by `/#/` in order to express the
+routing data as a URL fragment and maximise privacy.  For instance, a
+matrix-URI aware browser with no registered matrix URI handler would convert
+`matrix:r/fosdem:fosdem.org?default_client=https://chat.fosdem.org` to
+`https://chat.fosdem.org/#/r/fosdem:fosdem.org`.  It's then up to that target
+webapp to springboard the user into the appropriate web/mobile/native app.  If
+no `default_client` parameter is specified, the value defaults to
+`https://matrix.to` (which will need to be separately modified to support the
+MSC2312 URI scheme).
+
+* `default_hs` optionally specifies the suggested server which new users
+following this link should use to connect to Matrix.  For instance, a link
+whose author doesn't care which client is used to connect, but recommends a
+specific homeserver might be:
+`matrix:r/general:mozilla.org?default_hs=mozilla.org`.  In the absence of any
+registered matrix URI handler, this would be converted to
+`https://matrix.to/#/r/general:mozilla.org?default_hs=mozilla.org` (as
+`default_client` defaults to `https://matrix.to`).  The springboard webapp
+(matrix.to in this instance) then has to pass the parameter through to the
+actual client for use in login/registration.
```

Comment:
default_hs and default_client is specifically for initial onboarding. neither should be used if the client already has an account already.

Reply:
And that's what I'm aiming at: if they are intended to be used in conjunction why do we provide two parameters instead of one? Wouldn't `default_client` be enough, and even further: if these metadata are specifically for onboarding, why don't we name the parameter by the function - `onboard_at=encodedURL` that includes everything necessary in it?